napari_plugin_devtools/validation.py

```diff
@@ -0,0 +1,126 @@
+import inspect
+import json
+import os
+from types import FunctionType
+from typing import Callable, Dict, List, Tuple, Union
+
+from napari.plugins import plugin_manager
+from napari_plugin_engine import HookImplementation
+from pkginfo import SDist, Wheel
+
+functions: Dict[Tuple[str, str], Callable] = dict()
+
+
+def validate_artifacts(folder):
+    """Validate the artifacts built for distribution.
+
+    Parameters
+    ----------
+    folder : str
+        artifacts folder after build with setuptools
+
+    Raises
+    ------
+    AssertionError
+        If the artifact is not build properly,
+    """
+    for pkgpath in os.listdir(folder):
+        pkgpath = os.path.join(folder, pkgpath)
+        if pkgpath.endswith('.tar.gz'):
+            dist = SDist(pkgpath)
+        elif pkgpath.endswith('.whl'):
+            dist = Wheel(pkgpath)
+        else:
+            print(f'Not a valid format {pkgpath}')
+            continue
+        assert 'Framework :: napari' in dist.classifiers
+        print(f'validated {pkgpath}')
+
+
+def validate_function(
+    args: Union[Callable, List[Callable]],
+    hookimpl: HookImplementation,
+):
+    """Validate the function is properly implemented in napari supported format
+
+    Parameters
+    ----------
+    args : Union[Callable, List[Callable]]
+        function to validate
+    hookimpl : HookImplementation
+        implementation of the hook
+
+    Raises
+    ------
+    AssertionError
+        When function is not in supported format
+    """
+    plugin_name = hookimpl.plugin_name
+    hook_name = '`napari_experimental_provide_function`'
+
+    assert plugin_name is not None, "No plugin name specified"
+    for func in args if isinstance(args, list) else [args]:
+        assert not isinstance(func, tuple), (
+            "To provide multiple function widgets "
```

Comment:
We should now just call these `functions` and not `function widgets` since we renamed the hook impl